pkg/shares/share_merging.go

```diff
@@ -132,3 +134,43 @@ func ParseMsgs(shares [][]byte) (coretypes.Messages, error) {
 		MessagesList: msgList,
 	}, nil
 }
+
+// ShareSequence represents a contiguous sequence of shares that are part of the
+// same namespace and message. For compact shares, one share sequence exists per
+// reserved namespace. For sparse shares, one share sequence exists per message.
+type ShareSequence struct {
+	NamespaceID namespace.ID
+	Shares      []Share
+}
+
+func ParseShares(rawShares [][]byte) (result []ShareSequence, err error) {
+	currentSequence := ShareSequence{}
+
+	for _, rawShare := range rawShares {
+		share, err := NewShare(rawShare)
+		if err != nil {
+			return result, err
+		}
+		infoByte, err := share.InfoByte()
+		if err != nil {
+			return result, err
+		}
+		if infoByte.IsMessageStart() {
+			if len(currentSequence.Shares) > 0 {
+				result = append(result, currentSequence)
+			}
+			currentSequence = ShareSequence{
+				Shares:      []Share{share},
+				NamespaceID: share.NamespaceID(),
+			}
+		}
+		if !infoByte.IsMessageStart() {
+			if !bytes.Equal(currentSequence.NamespaceID, share.NamespaceID()) {
```

Comment:
[comment for other reviewers/posterity]
this isn't actually consensus breaking I don't think, even tho we weren't previously checking for this, because is order for this to be hit, the incorrect info byte has to be used which won't happen due to validators inserting it themselves and comparing the data root.

Reply:
on second thought, once we start using this, it will be consensus breaking with a dishonest majority

Reply:
> once we start using this

I thought `ParseShares` was for celestia-node consumption. Do we expect the state machine to invoke this API?  I didn't expect this API to be consensus breaking because I assumed the state machine would not call it.

Reply:
even if light clients are the only ones rejecting a block, then that is still consensus breaking (just like a change to fraud proofs)

Reply:
at least, I think its consensus breaking... :sweat_smile: 

imo, it makes sense to eventually use this code when we merge shares. It keeps things consistent and can possibly even inform other implementations on how to use universal shares.